src/Common/src/TypeSystem/Interop/IL/Marshaller.cs

```diff
@@ -1610,6 +1610,29 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
 
     class SafeHandleMarshaller : ReferenceMarshaller
     {
+        protected override void AllocNativeToManaged(ILCodeStream codeStream)
+        {
+            var ctor = ManagedType.GetParameterlessConstructor();
+            if (ctor == null)
+            {
+                var emitter = _ilCodeStreams.Emitter;
+
+                MethodSignature ctorSignature = new MethodSignature(0, 0, Context.GetWellKnownType(WellKnownType.Void),
+                      new TypeDesc[] { Context.GetWellKnownType(WellKnownType.String) });
+                MethodDesc exceptionCtor = InteropTypes.GetMissingMemberException(Context).GetKnownMethod(".ctor", ctorSignature);
+
+                MetadataType type = ManagedType as MetadataType;
+                string name = type != null ?  type.Name : "Subclass of SafeHandle";
+                codeStream.Emit(ILOpcode.ldstr, emitter.NewToken(String.Format("'{0}' does not have a default constructor. Subclasses of SafeHandle must have a default constructor to support marshaling a Windows HANDLE into managed code.", name)));
```

Comment:
This should be a helper that calls Resource.Format in S.P.Interop and supply the typehandle/name.